server/controllers/devices.coffee

```diff
@@ -14,13 +43,188 @@ extractCredentials = (header) ->
-module.exports.management = (req, res, next) ->
+# Check if device with <login> already exists
+deviceExists = (login, cb) ->
+    clientDS.post "request/device/byLogin/", key: login, (err, result, body) ->
+        if err
+            cb err
+        else if body.length is 0
+            cb null, false
+        else
+            cb null, body[0]
+
+initAuth = (req, cb) ->
+    # Authenticate the request
+    [username, password] = extractCredentials req.headers['authorization']
+    # Initialize user
+    user = {}
+    user.body = password: password
+    req.headers['authorization'] = undefined
+    cb user
+
+## Controller actions
+
+module.exports.create = (req, res, next) ->
+    # Create device :
+    #       * create device document
+    #       * create device access
+    createDevice = (device, cb) =>
+        device.docType = "Device"
+        # Create device document
+        clientDS.post "data/", device, (err, result, docInfo) ->
+            return cb(err) if err?
+            # Create access for this device
+            access =
+                login: device.login
+                password: randomString 32
+                app: docInfo._id
+                permissions: device.permissions or defaultPermissions
+            clientDS.post 'access/', access, (err, result, body) ->
+                return cb(err) if err?
+                data =
+                    password: access.password
+                    login: device.login
+                    permissions: access.permissions
+                # Return access to device
+                cb null, data
+
+    # Check if user is authenticated
+    authenticator = passport.authenticate 'local', (err, user) ->
+        if err
+            console.log err
+            next err
+        else if user is undefined or not user
+            error = new Error "Bad credentials"
+            error.status = 401
+            next error
+        else
+            # Check if name is correctly declared
+            device = req.body
+            if not device?.login?
+                error = new Error "Name isn't defined in req.body.login"
+                error.status = 400
+                next error
+            else
+                # Create device
+                device.docType = "Device"
+
+                # Check if an other device hasn't the same name
+                deviceExists device.login, (err, exist) ->
+                    if err
+                        next err
+                    else if exist
+                        error = new Error "This name is already used"
+                        error.status = 400
+                        next error
+                    else
+                        # Create device
+                        createDevice device, (err, data) ->
+                            if err?
+                                next err
+                            else
+                                res.send 201, data
+
+
+    initAuth req, (user) ->
```

Comment:
This whole module.exports.create function is hard to follow. May be you could split it several middlewares ?